cpp/log/cluster_state_controller-inl.h

```diff
@@ -4,20 +4,75 @@
 #include "log/cluster_state_controller.h"
 
 #include <functional>
+#include <stdint.h>
 
+#include "fetcher/peer.h"
 #include "log/database.h"
 #include "log/etcd_consistent_store-inl.h"
 #include "proto/ct.pb.h"
 
 namespace cert_trans {
 
-// TODO(alcutter): Need a better system for hanging tasks onto events, Pierre's
-//                 task-a-palooza idea perhaps?
+namespace {
+
+
+std::unique_ptr<AsyncLogClient> BuildAsyncLogClient(const std::shared_ptr<libevent::Base>& base, const ct::ClusterNodeState& state) {
+  CHECK(state.has_hostname());
+  CHECK(!state.hostname().empty());
+  CHECK_GT(state.log_port(), 0);
+  CHECK_LE(state.log_port(), UINT16_MAX);
+
+  return std::unique_ptr<AsyncLogClient>(
+      new AsyncLogClient(base, "http://" + state.hostname() + ":" +
+                                   std::to_string(state.log_port())));
+}
+
+
+}  // namespace
+
+
+template <class Logged>
+class ClusterStateController<Logged>::ClusterPeer : Peer {
+ public:
+  ClusterPeer(const std::shared_ptr<libevent::Base>& base,
+              const ct::ClusterNodeState& state)
+      : Peer(BuildAsyncLogClient(base, state)), state_(state) {
+  }
+
+  int64_t TreeSize() const override {
+    std::lock_guard<std::mutex> lock(lock_);
+    return state_.contiguous_tree_size();
```

Comment:
You probably want `newest_sth().tree_size()` here, 'cos `contiguous_tree_size` is about to disappear :)